### PR TITLE
Performance/28 aggregation speed

### PR DESCRIPTION
**Pull request by Nova-2119**
### Benefits

Speed benefits when using analytic aggregators. Example speed difference:
When running all centrality analytics and using the append aggregator:
1000/1000 Graph: 14,000 ms -> 300 ms
5000/5000 Graph: 525,475ms -> 442 ms

### Possible Drawbacks

Side effect: More results are now shown, previously when aggregating the result would not be shown if any of the analytic produced a null/0 result. Now they show as long as at least one analytic has a non null/0 result.

### Applicable Issues

https://github.com/constellation-app/constellation/issues/28

### CODE REVIEW

File: CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/aggregators/MaxScoreAggregator.java [hunk 1]
```diff
@@ -49,43 +46,16 @@ public ScoreResult aggregate(final List<ScoreResult> results) {
...
-                        for (final String scoreName : score.getNamedScores().keySet()) {
-                            isNull |= score.isNull();
-                            scores.add(score.getNamedScore(scoreName));
-                        }
-                    }
-                }
-            }
-
+        results.forEach(scoreResult -> combinedResults.combine(scoreResult));
+        combinedResults.getResult().forEach((IdentificationData key, ElementScore value) -> {
```
Comment (antares1470) [inline CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/aggregators/MaxScoreAggregator.java:50]: omit types from the lambda parameters if possible

File: CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/aggregators/MaxScoreAggregator.java [hunk 2]
```diff
@@ -49,43 +46,16 @@ public ScoreResult aggregate(final List<ScoreResult> results) {
...
             final Map<String, Float> aggregateScores = new HashMap<>();
-            aggregateScores.put(SCORE_NAME, scores.stream().reduce(Math::max).orElse((float) 0.0));
-
-            aggregateResult.add(new ElementScore(type, id, identifier, isNull, aggregateScores));
-        }
-
+            aggregateScores.put(SCORE_NAME, value.getNamedScores().values().stream().reduce(Math::max).orElse((float) 0.0));
```
Comment (antares1470) [inline CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/aggregators/MaxScoreAggregator.java:52]: noting you didn't add this but if you're able to change `(float) 0.0` to `0.0f`, that would be great

File: CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/aggregators/MeanScoreAggregator.java [hunk 3]
```diff
@@ -49,40 +46,13 @@ public ScoreResult aggregate(final List<ScoreResult> results) {
...
-                        for (final String scoreName : score.getNamedScores().keySet()) {
-                            isNull |= score.isNull();
-                            scores.add(score.getNamedScore(scoreName));
-                        }
-                    }
-                }
-            }
-
+        results.forEach(scoreResult -> combinedResults.combine(scoreResult));
+        combinedResults.getResult().forEach((IdentificationData key, ElementScore value) -> {
```
Comment (antares1470) [inline CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/aggregators/MeanScoreAggregator.java:50]: omit types from lambda parameters if possible

File: CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/aggregators/MeanScoreAggregator.java [hunk 4]
```diff
@@ -49,40 +46,13 @@ public ScoreResult aggregate(final List<ScoreResult> results) {
...
             final Map<String, Float> aggregateScores = new HashMap<>();
-            aggregateScores.put(SCORE_NAME, scores.stream().reduce((x, y) -> x + y).orElse((float) 0.0) / scores.size());
-
-            aggregateResult.add(new ElementScore(type, id, identifier, isNull, aggregateScores));
-        }
-
+            aggregateScores.put(SCORE_NAME, value.getNamedScores().values().stream().reduce((x,y) -> x+y).orElse((float) 0.0) / value.getNamedScores().size());
```
Comment (antares1470) [inline CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/aggregators/MeanScoreAggregator.java:52]: `0.0f`

File: CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/aggregators/MinScoreAggregator.java [hunk 5]
```diff
@@ -49,40 +46,13 @@ public ScoreResult aggregate(final List<ScoreResult> results) {
...
-                        for (final String scoreName : score.getNamedScores().keySet()) {
-                            isNull |= score.isNull();
-                            scores.add(score.getNamedScore(scoreName));
-                        }
-                    }
-                }
-            }
-
+        results.forEach(scoreResult -> combinedResults.combine(scoreResult));
+        combinedResults.getResult().forEach((IdentificationData key, ElementScore value) -> {
```
Comment (antares1470) [inline CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/aggregators/MinScoreAggregator.java:50]: omit lambda parameter types if possible

File: CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/aggregators/MinScoreAggregator.java [hunk 6]
```diff
@@ -49,40 +46,13 @@ public ScoreResult aggregate(final List<ScoreResult> results) {
...
             final Map<String, Float> aggregateScores = new HashMap<>();
-            aggregateScores.put(SCORE_NAME, scores.stream().reduce(Math::min).orElse((float) 0.0));
-
-            aggregateResult.add(new ElementScore(type, id, identifier, isNull, aggregateScores));
-        }
-
+            aggregateScores.put(SCORE_NAME, value.getNamedScores().values().stream().reduce(Math::min).orElse((float) 0.0));
```
Comment (antares1470) [inline CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/aggregators/MinScoreAggregator.java:52]: `0.0f`

File: CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/aggregators/SumScoreAggregator.java [hunk 7]
```diff
@@ -49,40 +46,13 @@ public ScoreResult aggregate(final List<ScoreResult> results) {
...
-                        for (final String scoreName : score.getNamedScores().keySet()) {
-                            isNull |= score.isNull();
-                            scores.add(score.getNamedScore(scoreName));
-                        }
-                    }
-                }
-            }
-
+        results.forEach(scoreResult -> combinedResults.combine(scoreResult));
+        combinedResults.getResult().forEach((IdentificationData key, ElementScore value) -> {
```
Comment (antares1470) [inline CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/aggregators/SumScoreAggregator.java:50]: omit parameter types if possible

File: CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/aggregators/SumScoreAggregator.java [hunk 8]
```diff
@@ -49,40 +46,13 @@ public ScoreResult aggregate(final List<ScoreResult> results) {
...
             final Map<String, Float> aggregateScores = new HashMap<>();
-            aggregateScores.put(SCORE_NAME, scores.stream().reduce((x, y) -> x + y).orElse((float) 0.0));
-
-            aggregateResult.add(new ElementScore(type, id, identifier, isNull, aggregateScores));
-        }
-
+            aggregateScores.put(SCORE_NAME, value.getNamedScores().values().stream().reduce((x,y) -> x+y).orElse((float) 0.0));
```
Comment (antares1470) [inline CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/aggregators/SumScoreAggregator.java:52]: `0.0f`

File: CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/results/IdentificationData.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    public int hashCode() {
+        int hash = 3;
+        hash = 71 * hash + Objects.hashCode(this.elementType);
+        hash = 71 * hash + this.elementId;
+        hash = 71 * hash + Objects.hashCode(this.identifier);
+        return hash;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```
Comment (antares1470) [inline CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/results/IdentificationData.java:77]: obj can be final

File: CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/results/ScoreResult.java [hunk 10]
```diff
@@ -57,6 +65,13 @@ public ElementScore(final GraphElementType elementType, final int elementId,
             super(elementType, elementId, identifier, isNull);
             this.namedScores = namedScores;
         }
+        
+        static public ElementScore combineReplace(ElementScore es1, ElementScore es2) {
```
Comment (antares1470) [inline CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/results/ScoreResult.java:69]: parameters can be final. Also change `static public` to `public static`

File: CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/results/ScoreResult.java [hunk 11]
```diff
@@ -57,6 +65,13 @@ public ElementScore(final GraphElementType elementType, final int elementId,
...
+            Map<String, Float> combinedNamedScores = new HashMap<>(es1.getNamedScores());
```
Comment (antares1470) [inline CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/results/ScoreResult.java:70]: final

**Review by antares1470 — CHANGES_REQUESTED** [review]

**Review by antares1470 — APPROVED** [review]

**Review by arcturus2 — APPROVED** [review]

**Pull request merged**